src/session.c

```diff
@@ -0,0 +1,449 @@
+        }
+        if ((slots[i].token.flags & CKF_TOKEN_INITIALIZED) == 0) {
+            continue;
+        }
+
+        id = slots[i].id;
+        slot = &slots[i];
+        ret = CKR_OK;
+
+        if (uri) {
+            CK_TOKEN_INFO token = slot->token;
+
+            /* skip slots that do not match */
+            ret = p11prov_uri_match_token(uri, &token);
+            if (ret == CKR_CANCEL) {
+                continue;
+            }
+            if (ret == CKR_OK && (token.flags & CKF_LOGIN_REQUIRED)) {
+                pthread_mutex_lock(&slot->sessions->lock);
+                ret = token_login(provctx, slot, uri, pw_cb, pw_cbarg);
+                pthread_mutex_unlock(&slot->sessions->lock);
+            }
+        }
+        break;
+    }
+
+    if (ret == CKR_OK) {
+        /* Found a slot, return it and the next slot to the caller for
+         * continuation if the current slot does not yield the desired
+         * results */
+        *slotid = id;
+        if (next_slotid) {
+            if (i + 1 < nslots) {
+                *next_slotid = slots[i + 1].id;
+            } else {
+                *next_slotid = CK_UNAVAILABLE_INFORMATION;
+            }
+        }
+    } else {
+        *next_slotid = CK_UNAVAILABLE_INFORMATION;
+    }
+
+    if (ret != CKR_OK) {
+        return ret;
+    }
+
+    /* acquire lock */
+    pthread_mutex_lock(&slot->sessions->lock);
+
+    if (slot->sessions->free_sessions) {
+        sess = slot->sessions->free_sessions;
+        LIST_REMOVE(slot->sessions->free_sessions, sess);
+        check_session = true;
+    } else {
+        sess = p11prov_session_new(provctx, slot);
+        if (sess == NULL) {
+            ret = CKR_HOST_MEMORY;
+            goto unlock;
+        }
+    }
+    if (sess->session == CK_INVALID_HANDLE) {
+        ret = p11prov_session_open(sess, false, NULL, 0);
+        if (ret != CKR_OK) {
+            goto unlock;
+        }
+        check_session = false;
+    }
+    LIST_INSERT(slot->sessions->busy_sessions, sess);
+
+unlock:
+    /* free lock */
+    pthread_mutex_unlock(&slot->sessions->lock);
+    if (ret != CKR_OK) {
+        p11prov_session_free(sess);
+        return ret;
+    }
+
+    if (check_session == true) {
+        /* check that session is still ok */
+        CK_FUNCTION_LIST *f;
+        CK_SESSION_INFO session_info;
+
+        f = p11prov_ctx_fns(provctx);
+        if (f == NULL) {
+            return CKR_GENERAL_ERROR;
+        }
+
+        ret = f->C_GetSessionInfo(sess->session, &session_info);
+        switch (ret) {
+        case CKR_OK:
+            break;
+        case CKR_SESSION_CLOSED:
+        case CKR_SESSION_HANDLE_INVALID:
+            /* try to open a session again */
+            sess->session = CK_INVALID_HANDLE;
+            ret = p11prov_session_open(sess, false, NULL, 0);
+            if (ret == CKR_OK) {
+                pthread_mutex_lock(&slot->sessions->lock);
+                LIST_INSERT(slot->sessions->busy_sessions, sess);
```

Comment:
Good point, I do not particularly like this whole list management, I wonder if there is a better way...

Reply:
worse that that, on failure it still returns the session to the caller without an error ...

Reply:
However I do not see how 2 happens, given that in 2, if we do open a session we also set check_session to false. And w/o 2 there is no problem with LIST_INSERT() ...

Reply:
on error paths we may also "lose" sessions as they are removed from the free_sessions list and not added back.

Reply:
I really do not like this 2 lists business, I am going to change this whole thing using a free/busy int on the session and a linear list scan I think. It may be slightly less efficient when searching for a free session from time to time, but manipulations will not need locks, as a simple atomic xompare_exchange operation can be used to flip a session from free to busy and back.

Reply:
This happens when trying to reuse an existing session handle, which has been closed/invalidated by the pkcs11 library for some reason.